### PR TITLE
Update docs for split stdout / stderr Pod logging

### PR DESCRIPTION
**Pull request by knight42**
This is a ~placeholder~ PR for the new feature https://github.com/kubernetes/enhancements/issues/3288. I will update the document after the implemention PR get merged.

KEP issue: https://github.com/kubernetes/enhancements/issues/3288
Code (k/k) update PR(s):
* https://github.com/kubernetes/kubernetes/pull/127360
* https://github.com/kubernetes/kubernetes/pull/128159

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+_build:
+  list: never
+  render: false
+
+stages:
+  - stage: alpha
+    defaultValue: false
+    fromVersion: "1.32"
+---
+Allow retrieving either stdout or stderr log stream of the given container.
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md:13]: Try this:
```suggestion
Enable fetching specific log streams (either stdout or stderr) from a container's log streams, using the Pod API.
```

**Review by sftim — COMMENTED** [review]
Ideally, also update https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#examine-pod-logs

**Comment by spurin** [comment]
Hello @knight42 :wave: please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday November 19th 2024 18:00 PST. Thank you!

**Comment by tengqm** [comment]
Looks like the upstream code change has landed.
/lgtm

**Comment by spurin** [comment]
Hello @knight42 :wave:, 1.32 Docs Shadow here,

A quick reminder to get your PR ready for review before the deadline, Tuesday November 19th at 18:00 PDT. For additional information, refer to: [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review).

Thank you!

**Comment by haircommander** [comment]
@knight42 is this still a draft or is this ready?

**Comment by knight42** [comment]
> Ideally, also update [kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod#examine-pod-logs](https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#examine-pod-logs)

Unfortunately the kubectl PR https://github.com/kubernetes/kubernetes/pull/128159 has not been merged yet, I think we'd better update this document later?

**Comment by tengqm** [comment]
> > Ideally, also update [kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod#examine-pod-logs](https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#examine-pod-logs)
> 
> Unfortunately the kubectl PR [kubernetes/kubernetes#128159](https://github.com/kubernetes/kubernetes/pull/128159) has not been merged yet, I think we'd better update this document later?

Yup. We'd better leave this a WIP and see if the upstream PR can be merged.

**Comment by sftim** [comment]
/retitle Update docs for split stdout / stderr Pod logging

File: content/en/docs/concepts/cluster-administration/logging.md [hunk 2]
```diff
@@ -75,6 +75,34 @@ appending a container name to the command, with a `-c` flag, like so:
 kubectl logs counter -c count
 ```
 
+Since v1.32, you can request a specific log stream of a container in a running Pod.
+This functionality requires the
+`PodLogsQuerySplitStreams` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to be enabled.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/logging.md:80]:
```suggestion

### Container log streams

{{< feature-state feature_gate_name="PodLogsQuerySplitStreams" >}}

As an alpha feature, the kubelet can split out the logs from the two standard streams produced
by a container: [standard output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout))
and [standard error](https://en.wikipedia.org/wiki/Standard_streams#Standard_error_(stderr)).
To use this behavior, you must enable the `PodLogsQuerySplitStreams`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
```

File: content/en/docs/concepts/cluster-administration/logging.md [hunk 3]
```diff
@@ -75,6 +75,34 @@ appending a container name to the command, with a `-c` flag, like so:
...
+Unfortunately, only the server side changes have been merged so far; the [corresponding kubectl changes](https://github.com/kubernetes/kubernetes/pull/128159) are still under review.
+In the meantime, you can still use this feature by directly accessing the API.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/logging.md:82]: We prefer to write documentation that is more [timeless](https://developers.google.com/style/timeless-documentation). Try this:
```suggestion
With that feature gate enabled, Kubernetes {{< skew currentVersion >}} allows access to these
log streams directly via the Pod API. You can fetch a specific stream by specifying the stream name (either `Stdout` or `Stderr`),
using the `stream` query string. You must have access to read the `log` subresource of that Pod.
```

File: content/en/docs/concepts/cluster-administration/logging.md [hunk 4]
```diff
@@ -75,6 +75,34 @@ appending a container name to the command, with a `-c` flag, like so:
...
+
+```shell
+kubectl get --raw "/api/v1/namespaces/default/pods/counter-err/log?stream=Stderr"
+```
+
+Once kubectl is updated, you will be able to easily retrieve a specific log stream using:
+
+```shell
+kubectl logs --stream=stderr counter-err
+```
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/logging.md:104]: We [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in docs, so:
```suggestion
```

File: content/en/docs/tasks/debug/debug-application/debug-running-pod.md [hunk 5]
```diff
@@ -345,6 +345,21 @@ If your container has previously crashed, you can access the previous container'
...
+
+```shell
+kubectl get --raw "/api/v1/namespaces/default/pods/${POD_NAME}/log?container=${CONTAINER_NAME}&stream=Stderr"
+```
+
+Once kubectl is updated, you will be able to easily retrieve a specific log stream using:
+
+```shell
+kubectl logs --stream=stderr ${POD_NAME} ${CONTAINER_NAME}
+```
```
Comment (sftim) [inline content/en/docs/tasks/debug/debug-application/debug-running-pod.md:361]: Until there's more support, this advice isn't much use to a reader who wants to troubleshoot a Pod.
```suggestion
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md:1]: Is this right?

**Review by sftim — CHANGES_REQUESTED** [review]
Why does this PR add _two_ feature gates?

**Comment by sftim** [comment]
Given the question at https://github.com/kubernetes/website/pull/48405#pullrequestreview-2468866590

/lgtm cancel

/hold

OK to unhold once that question is addressed.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md
Comment (knight42) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/split-stdout-stderr.md:1]: oops... let me remove this file

**Review by sftim — COMMENTED** [review]
LGTM

**Comment by knight42** [comment]
> Why does this PR add two feature gates?

Ah, I think one was submitted by mistake, I’ve deleted the incorrect one.

**Comment by sftim** [comment]
/lgtm
/approve

/hold
You can either:
- squash this to 1 commit (keep the same parent commit, 35e7e1d00c0f15c3124b27f309ddc4ea5764b71c, though)
  and then unhold
- unhold as-is

First option is slightly tidier though.

**Pull request merged**